codecs/src/aper/encode/mod.rs

```diff
@@ -48,11 +48,13 @@ pub fn encode_sequence_header(
         ));
     }
 
+    let mut offset = 0;
     if is_extensible {
-        data.encode_bool(extended);
+        data.set_bit(0, extended); // FIXME: Correct?
+        offset += 1;
     }
 
-    data.append_bits(optionals);
+    data.swap_bits(optionals, offset);
     Ok(())
 }
```

Comment:
@nplrkn : These are the changes to the APIs as I have implemented the `derive_macros`. It takes your original idea, but instead of `append`ing, we now have to `swap` bits. Hence the changes. The generated structure and code looks something like the following.

```rust

// Test Structures for encoding sequences. `codecs_derive/tests/08-seq.rs`

#![allow(non_camel_case_types, dead_code)]

use asn1_codecs_derive::AperCodec;

#[derive(Debug, AperCodec)]
#[asn(type = "INTEGER", lb = "0", ub = "65535")]
pub struct ProtocolIE_ID(u16);

#[derive(Debug, AperCodec)]
#[asn(type = "ENUMERATED", lb = "0", ub = "2")]
pub struct Criticality(u8);
impl Criticality {
    const REJECT: u8 = 0u8;
    const IGNORE: u8 = 1u8;
    const NOTIFY: u8 = 2u8;
}

#[derive(Debug, AperCodec)]
#[asn(type = "SEQUENCE", extensible = false, optional_fields = 1)]
pub struct Bearers_SubjectToEarlyStatusTransferListItem {
    pub id: ProtocolIE_ID,

    #[asn(optional_idx = 0)]
    pub critcality: Option<Criticality>,
}

// Generated encode  -- `cargo expand --test 08-seq` 
fn encode(
        &self,
        data: &mut asn1_codecs::aper::AperCodecData,
    ) -> Result<(), asn1_codecs::aper::AperCodecError> {
        let mut header_bits = 1;
        if false {
            header_bits += 1;
        }
        data.reserve(header_bits as usize);
        let mut bitmap = ::bitvec::vec::BitVec::<bitvec::prelude::Msb0, u8>::repeat(0 != 0, 1);
        self.id.encode(data)?;
        if self.critcality.is_some() {
            let critcality = self.critcality.as_ref().unwrap();
            bitmap.set(0, true);
            critcality.encode(data)?;
        } else {
        }
        data.seek(0_usize);
        asn1_codecs::aper::encode::encode_sequence_header(data, false, &mut bitmap, false)?;
        Ok(())
    }
```

Reply:
Let me know if it's okay to make this change to `encode_seq_header` API.

Reply:
Latest on this @nplrkn : I don't think I need to make changes to `encode_sequence_header`. I have implemented `derive` macro for SEQUENCE using your original idea.

I realized with proposed API changes the encode functionality might be slower and was little error prone (rather than simply extending a bit buffer.) 

Semantically, i would still prefer to `encode_sequence_header` to happen in the end, but may be that can be visited some other time.

Also, I have left intact the added APIs to `AperCodecData` at some point of time they may be useful. But for now - I am not deleting them.

With this - I believe my original question is answered. 

I can merge this right?